In [2]:
import numpy as np
import pandas as pd
import ISLP

We will now consider the Boston housing data set, from the ISLP
library.

(a) Based on this data set, provide an estimate for the population
mean of medv. Call this estimate $\hat{\mu}$

In [3]:
df = ISLP.load_data('Boston')
print(df.shape)
print(df.describe())
mu_hat = df['medv'].mean()
print(mu_hat)

(506, 13)
             crim          zn       indus        chas         nox          rm  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              age         dis         rad         tax     ptratio       lstat  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534   12.653063   
std     28.148861

(b) Provide an estimate of the standard error of $\hat{\mu}$. Interpret this
result.
Hint: We can compute the standard error of the sample mean by
dividing the sample standard deviation by the square root of the
number of observations.

In [4]:
n = df.shape[0]
s = df['medv'].std()
std_err = s / np.sqrt(n)
print(std_err)

0.4088611474975351


The standard error of $\hat{\mu}$ is approximately $0.4085$ (in thousands of dollars). This value is the estimated standard deviation of the sample mean, quantifying the average amount the sample mean would vary from the true population mean if the sampling process were repeated many times. Since the value is small, it suggests that our estimate $\hat{\mu}$ is highly precise.

(c) Now estimate the standard error of $\hat{\mu}$ using the bootstrap. How
does this compare to your answer from (b)?

In [5]:
B = 1000000
np.random.seed(42)
mu_hat_bootstrap_samples = []
medv_data = df['medv'].values

for i in range(B):
    # 使用 np.random.choice 進行有放回的隨機抽樣
    # replace=True 執行有放回抽樣
    bootstrap_sample = np.random.choice(medv_data, size=n, replace=True)
    
    # 計算並儲存 bootstrap 樣本的平均數
    mu_hat_bootstrap_samples.append(np.mean(bootstrap_sample))

# 計算 Bootstrap 樣本平均數的標準差，即為 SE($\hat{\mu}$)
SE_mu_hat_bootstrap = np.std(mu_hat_bootstrap_samples, ddof=1) # 使用 ddof=1 計算樣本標準差

print(SE_mu_hat_bootstrap)

0.40871082849660334


When the number of samples are bigger, the outcome is more close to the answer from (b). 

(d) Based on your bootstrap estimate from (c), provide a 95 % confidence interval for the mean of medv. Compare it to the results
obtained by using Boston['medv'].std() and the two standard error rule (3.9).
Hint: You can approximate a 95 % confidence interval using the
formula [$\hat{\mu}$− 2SE($\hat{\mu}$),$\hat{\mu}$ + 2SE($\hat{\mu}$)].

In [6]:
boost_start = mu_hat - 2 * SE_mu_hat_bootstrap
boost_end = mu_hat + 2 * SE_mu_hat_bootstrap

std_start = mu_hat - 2 * std_err
std_end = mu_hat + 2 * std_err

print(f"boostrap: [ {boost_start}, {boost_end} ]")
print(f"std: [ {std_start}, {std_end} ]")

boostrap: [ 21.71538466711747, 23.350227981103885 ]
std: [ 21.715084029115605, 23.35052861910575 ]


Two interval are almost the same, the high degree of agreement between the two methods confirms the reliability of the standard error estimate. For a large sample size ($n=506$), the Central Limit Theorem holds, making the simple formula (Std) a highly accurate approximation of the sampling distribution's variability.

(e) Based on this data set, provide an estimate, $\hat{\mu_{med}}$, for the median
value of medv in the population.

In [7]:
mu_hat_med = df['medv'].median()
print(mu_hat_med)

21.2


(f) We now would like to estimate the standard error of $\hat{\mu_{med}}$. Unfortunately, there is no simple formula for computing the standard
error of the median. Instead, estimate the standard error of the
median using the bootstrap. Comment on your findings.

In [9]:
B = 1000000
np.random.seed(42)
mu_hat_med_bootstrap_samples = []
medv_data = df['medv'].values


for i in range(B):
    # 使用 np.random.choice 進行有放回的隨機抽樣
    # replace=True 執行有放回抽樣
    bootstrap_sample = np.random.choice(medv_data, size=n, replace=True)
    
    # 計算並儲存 bootstrap 樣本的平均數
    mu_hat_med_bootstrap_samples.append(np.median(bootstrap_sample))

# 計算 Bootstrap 樣本平均數的標準差，即為 SE($\hat{\mu}$)
SE_mu_hat_med_bootstrap = np.std(mu_hat_med_bootstrap_samples, ddof=1) # 使用 ddof=1 計算樣本標準差

print(SE_mu_hat_med_bootstrap)
print(df['medv'].std())

0.378253178207486
9.19710408737982


The results show that the standard error of the median is relatively small compared to the standard deviation of the original data, reflecting the fact that the median is less sensitive to extreme values than the mean. The bootstrap approach gives a practical and reliable way to quantify the variability of the median.

(g) Based on this data set, provide an estimate for the tenth percentile of medv in Boston census tracts. Call this quantity $\hat{\mu_{0.1}}$ .
(You can use the np.percentile() function.) 

In [12]:
mu_01 = np.percentile(df['medv'], 10)
mu_01

np.float64(12.75)

(h) Use the bootstrap to estimate the standard error of $\hat{\mu_{0.1}}$. Comment on your findings.

In [13]:
boot_p10 = []
for _ in range(B):
    sample = np.random.choice(medv_data, size=n, replace=True)
    boot_p10.append(np.percentile(sample, 10))

boot_se_p10 = np.std(boot_p10, ddof=1)
boot_se_p10


np.float64(0.5027402622767361)

Same as the median, the standard error shows that the tenth percentile also less senesitive to extreme values than the mean.